binary_gentoo/internal/cli/tree_diff.py

```diff
@@ -3,12 +3,72 @@
 
 import filecmp
 import os
+import re
 import sys
 from argparse import ArgumentParser
+from typing import Iterable, Set
 
-from ..reporter import exception_reporting
+from ..reporter import announce_and_check_output, exception_reporting
 from ._parser import add_version_argument_to
 
+_keywords_pattern = re.compile('KEYWORDS="(?P<keywords>[^"]*)"')
+
+
+def _get_relevant_keywords_set_for(ebuild_content: str,
+                                   accept_keywords: Iterable[str]) -> Set[str]:
+    match = _keywords_pattern.search(ebuild_content)
+    if match is None:
+        # if the KEYWORDS variable is not found in the ebuild,
+        # we will assume the ebuild needs to be listed
+        ebuild_keywords = accept_keywords
```

Comment:
Yes, in fact that is right. I investigated a bit closer the problems I had that led to putting it in like this, and I must admit I was too quick in my analysis. I'll dig a bit deeper and see if I need to make additional modifications.

Reply:
In fact, the current code will also examine files such as `app-admin/calamares/files/calamares-3.2.32.1-remove-broken-tests.ebuild`. If that line matches all the criteria (not changed, includes KEYWORDS, etc.) it will also be outputted by `gentoo-tree-diff`. With the additional checks on KEYWORDS, it's going to be less likely that it is outputted. I'm going to leave that as it is for now, because this behaviour was present before this PR already. Best to have a separate PR to fix that bug -- if it is important enough to be fixed.

Reply:
> We can either look for inherit lines for those eclasses or put ebuilds in categories `acct-group` and `acct-user` in an always-matches list, but other than that mismatch should be the fallback behavior. No?

Since the acct-group and acct-user eclasses don't contain actual code, I think it's okay if they are never outputted by `gentoo-tree-diff`.

Reply:
> In fact, the current code will also examine files such as `app-admin/calamares/files/calamares-3.2.32.1-remove-broken-tests.ebuild`. If that line matches all the criteria (not changed, includes KEYWORDS, etc.) it will also be outputted by `gentoo-tree-diff`. With the additional checks on KEYWORDS, it's going to be less likely that it is outputted. I'm going to leave that as it is for now, because this behaviour was present before this PR already. Best to have a separate PR to fix that bug -- if it is important enough to be fixed.

Intersting catch. I think it's the only such case…

```console
# find /usr/portage/ -mindepth 4  -type f -name \*.ebuild 2>/dev/null 
/usr/portage/app-admin/calamares/files/calamares-3.2.32.1-remove-broken-tests.ebuild
```

…and hence maybe is best fixed in Gentoo itself. I have down that [just now](https://github.com/gentoo/gentoo/commit/4e7137845d4e053f2143446cbd427add660f61de). If we want to be more robust here, line `package = category_plus_package.split(os.sep)[-1]` might be a good place to change.

> > We can either look for inherit lines for those eclasses or put ebuilds in categories `acct-group` and `acct-user` in an always-matches list, but other than that mismatch should be the fallback behavior. No?
> 
> Since the acct-group and acct-user eclasses don't contain actual code, I think it's okay if they are never outputted by `gentoo-tree-diff`.

Given that `gentoo-build` will still produce binary packages for these (when pulled in as a dependency), I guess that's still convenient in practice, then.